doc/compiler/debug/Loggers.md

```diff
@@ -0,0 +1,176 @@
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] https://openjdk.org/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+-->
+
+# Loggers
+
+A *Logger* is a simple structure for performing diagnostic I/O behind a
+lightweight API. There are several advantages to using a logging mechanism:
+
+- a Logger provides a consistent API to be used for all diagnostic output.
+  Historical Testarossa implementations employed multiple solutions
+  simultaneously with no standardization or consistency.
+
+- it will be easier to make broad changes in the future affecting all
+  diagnostic I/O if a single API is used. For example, introducing log levels
+  controlling diagnostic behaviour commonly used in logging infrastructure
+  found in other projects.
+
+- Loggers enable creative debugging practices. For example, a Logger need not
+  write to a file but be streamed out a socket instead, and it is
+  straightforward to capture diagnostic output at different parts of the
+  compilation process to different Loggers.
+
+- a Logger hides its implementation details behind a consistent, abstract API.
+  For example, a Logger could be implemented to use C stdio or Testarossa's
+  `TR::IO` for I/O, but the API for the developer remains the same (only the
+  installed Logger implementation is different).
+
+A `TR::Logger` is an abstract base class that declares a logging API. A family
+of useful Logger implementations is provided in the OMR compiler, but it is
+straightfoward for a developer to extend or create custom Loggers.
+
+## Logger API Functions
+
+The Logger API has a family of output functions available for use depending on
+the circumstance. For instance, not all output requires format specifiers, and
+in those circumstances API functions are available to output a constant string
+or even a single character. Such variants are more efficient because they are
+not scanned for format specifiers prior to output. Developers should choose the
+most appropriate and efficient function for their output needs.
+
+## Logger Implementations
+
+### NullLogger
+
+This is the simplest Logger that simply eats anything passed into it without
+outputting it anywhere. This is useful to install as a default Logger to catch
+any unguarded Logger output in a production setting. Knowing that some Logger
+will always be available means that code does not always need to check for the
+presence of a Logger before outputting.
+
+### CStdIOStreamLogger
+
+A Logger that outputs directly to a file using C stdio output functions.
+
+### TRIOStreamLogger
+
+A Logger that outputs directly to a file using Testarossa `TR::IO` output
+functions.
```

Comment:
I've looked very little at the underlying implementation of the old logging facilities, so I don't really have a sense for how the `TR::IO` output functions differ from the C stdio output functions, and I suspect other readers might be in the same boat.  Can you describe at a high-level this differs from `CStdIOStreamLogger` or point the reader to some relevant documentation for `TR::IO` that would make the difference clear, if any exists?